Address HSPF2 calculation problems for multispeed DX system

Nigusse opened the pull request:
Pull request overview
---------------------
 - Fixes #9909 
 - Corrects the HSPF2 calculation procedure for multi-speed DX heating coils to match the 2023 AHRI standards 210/240
 - HSPF2 value diffs are expected 
- Fix #10651



### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Nigusse committed:
Modified HSPF2 calculation

Nigusse committed:
Modified a fail unit test

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6469,18 +6469,30 @@ namespace StandardRatings {
+                        q_full = q_H3_full + (q_H1_full - q_H3_full) * t_ratio;
+                        // Equation 11.200 AHRI-2023
+                        p_full = p_H3_full + (p_H1_full - p_H3_full) * t_ratio;
                     }
```
The current code for Case III looks incorrect based on the review of the AHRI 2023 (2020) Standard for Performance Rating of Unitary Air-conditioning & Air-source Heat Pump Equipment. This code change matches the AHRI standard and corrects the logic flaws.

Myoldmopar replied:
Thanks @Nigusse

Nigusse commented on tst/EnergyPlus/unit/DXCoils.unit.cc:
```diff
@@ -4656,7 +4656,7 @@ TEST_F(EnergyPlusFixture, TestMultiSpeedCoilsAutoSizingOutput)
  DX Heating Coil Standard Rating Information, Coil:Heating:DX:MultiSpeed, ASHP HTG COIL, 34415.4, 20666.4, 6.56, 4
 ! <DX Heating Coil AHRI 2023 Standard Rating Information>, Component Type, Component Name, High Temperature Heating (net) Rating Capacity {W}, Low Temperature Heating (net) Rating Capacity {W}, HSPF2 {Btu/W-h}, Region Number
- DX Heating Coil AHRI 2023 Standard Rating Information, Coil:Heating:DX:MultiSpeed, ASHP HTG COIL, 34697.2, 20948.1, 5.34, 4
+ DX Heating Coil AHRI 2023 Standard Rating Information, Coil:Heating:DX:MultiSpeed, ASHP HTG COIL, 34697.2, 20948.1, 5.35, 4
```
Modified a failed unit test.

Nigusse reviewed (commented):
Code change walkthrough:
- the code change now match the 2023 AHRI Standard

Nigusse committed:
Added new unit tests

Nigusse committed:
Set the full capacity and energy at the next higher speed level

Nigusse committed:
Modified the equations for capacity and energy at low speed

Nigusse committed:
Modified the equations for capacity and energy at high speed

Nigusse committed:
allow HSPF2 calculation with flat curve

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6332,17 +6332,17 @@ namespace StandardRatings {
+                Real64 q_H1_full = Q_H1_Full(spnum + 1);
+                Real64 q_H2_full = Q_H2_Full(spnum + 1);
+                Real64 q_H3_full = Q_H3_Full(spnum + 1);
+                Real64 q_H4_full = Q_H4_Full(spnum + 1);
```
The full heating capacity must be evaluated at the next available speed, not at the current speed. It should be different from the low speed.

Myoldmopar replied:
It makes sense, and I believe you, but it would be great to get a second opinion on it.  Anyone have thoughts on this?

rraustad replied:
I don't know the standard but what will these values be used for? i.e., these would be used in some equation, so show the equation to support this change. For example, q_H1_full looks to be the capacity at speed 1 so why would we use speed 2 capacity?

Nigusse replied:
These capacities (H1, 2, 3, 4) are used to estimate the full capacity at a specified outdoor a condition through interpolation. The full capacity is then used to bracket the `building load` to get a speed level of the machine that matches the `building load`.

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6351,17 +6351,17 @@ namespace StandardRatings {
+                Real64 p_H1_full = P_H1_Full(spnum + 1);
+                Real64 p_H2_full = P_H2_Full(spnum + 1);
+                Real64 p_H3_full = P_H3_Full(spnum + 1);
+                Real64 p_H4_full = P_H4_Full(spnum + 1);
```
The full electric power input must be evaluated at the next available speed, not at the current speed. It should be different from the low speed.

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6332,17 +6332,17 @@ namespace StandardRatings {
 
                 // Equation 11.191 AHRI-2023
-                Real64 N_Hq = min(1.0, (q_H2_int - q_35_low) / (q_H2_full - q_35_low));
+                Real64 N_Hq = (q_H2_full != q_35_low) ? min(1.0, (q_H2_int - q_35_low) / (q_H2_full - q_35_low)) : 0.0;
```
Modified to allow evaluation of the HSPF2 with flat heating capacity modifier curves function of temperature.

jmarrec replied:
This fixes #10651


FWIW, and this looks like undefined behavior, but I was finding that in a release build with no floating point exceptions enabled, when you had the same Cap FT / Fflow curves, N_Hq was 0 but N_HE was 1.

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6351,17 +6351,17 @@ namespace StandardRatings {
 
                 // Equation 11.194 AHRI-2023
-                Real64 N_HE = min(1.0, (p_H2_int - p_35_low) / (p_H2_full - p_35_low));
+                Real64 N_HE = (p_H2_int != p_35_low) ? min(1.0, (p_H2_int - p_35_low) / (p_H2_full - p_35_low)) : 0.0;
```
It was modified to allow the evaluation of the HSPF2 with flat heating capacity modifier curves function of temperature.

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6375,35 +6375,63 @@ namespace StandardRatings {
+                    // equation 11.181
+                    q_low = Q_H2_Int(spnum);
+                    // equation 11.184
+                    p_low = P_H2_Int(spnum);
```
It was modified the equations used to calculate delivered heat and energy input at the `Low Speed` to match the 2023 AHRI Standard 210/240.

Nigusse commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -6375,35 +6375,63 @@ namespace StandardRatings {
+                    q_hs = Q_H3_Full(spnum + 1) + ((Q_H1_Full(spnum + 1) - Q_H3_Full(spnum + 1)) * ratio);
+                    // equation 11.200
+                    p_hs = P_H3_Full(spnum + 1) + ((P_H1_Full(spnum + 1) - P_H3_Full(spnum + 1)) * ratio);
                 }
```
Modified the equations used to calculate delivered heat and energy input at the `High Speed` to match the 2023 AHRI Standard 210/240.

rraustad replied:
OK, so here I see the equations. I still don't know if Q_H1_Full should be the "full" capacity at the speed being evaluated or the full capacity at the next highest speed. What in the standard prompted this change?

Nigusse replied:
My understanding is that the next higher speed level should be used to get the full speed capacity that bounds the current `building load`.

rraustad replied:
But if PLR at speed x is < 1 then the full capacity at that same speed would bound the current building load. So where in the standard does it say to use the next highest speed level?

Nigusse replied:
In the standard, different sections may apply to this code: one for a two-speed (two-capacity) machine and another one for a variable-speed machine. They are not an exact match for a discrete-speed machine with more than the two-speed. The HSPF2 calculation is based on the variable speed section of the standard, and it requires some interpretation to fit the discrete multispeed DX heating coil model we have in EnergyPlus. It requires some judgment on how to account for the difference between the EnergyPlus model and the variable speed machine procedure in the AHRI standard. Simply, it needs interpretation of the procedure for implementation in EnergyPlus. 

In a variable-speed machine, you have just one full speed, like the two-speed DX cooling coil model in EnergyPlus. In the multispeed DX heating coil, your full speed is a moving target, and depending on your load, it is the next available higher-speed capacity. Let's say a given load is bound between two-speed levels; the full load should point to the higher speed level, not the lower speed. This is my interpretation of how it should be. The current full capacity assignment was to the current speed, so I changed it to the next higher-speed capacity.

Nigusse commented on tst/EnergyPlus/unit/DXCoils.unit.cc:
```diff
@@ -4656,7 +4656,7 @@ TEST_F(EnergyPlusFixture, TestMultiSpeedCoilsAutoSizingOutput)
  DX Heating Coil Standard Rating Information, Coil:Heating:DX:MultiSpeed, ASHP HTG COIL, 34415.4, 20666.4, 6.56, 4
 ! <DX Heating Coil AHRI 2023 Standard Rating Information>, Component Type, Component Name, High Temperature Heating (net) Rating Capacity {W}, Low Temperature Heating (net) Rating Capacity {W}, HSPF2 {Btu/W-h}, Region Number
- DX Heating Coil AHRI 2023 Standard Rating Information, Coil:Heating:DX:MultiSpeed, ASHP HTG COIL, 34697.2, 20948.1, 5.34, 4
+ DX Heating Coil AHRI 2023 Standard Rating Information, Coil:Heating:DX:MultiSpeed, ASHP HTG COIL, 34697.2, 20948.1, 5.76, 4
```
Modified the HSPF2 values unit test to match latest results. Now the difference between HSPF and HSPF2 reduced to 12.2% from 18.6%.

Nigusse commented on tst/EnergyPlus/unit/DXCoils.unit.cc:
```diff
@@ -6880,4 +6880,773 @@ TEST_F(EnergyPlusFixture, Test_DHW_End_Use_Cat_Removal)
+    EXPECT_TRUE(compare_eio_stream(htg_coil_eio_output, true));
+}
+
 } // namespace EnergyPlus
```
Added three unit tests. These unit tests evaluate HSPF values for various compressor operation control minimum outdoor air temperature and fan power inputs. The unit tests demonstrate that the HSPF2 values are now closer to the HSPF values.

Nigusse reviewed (commented):
Code change walkthrough for additional issues identified while working on the original defect.

Nigusse commented:
See below comparison of the HSPF2 values of develop and this branch:

Develop: percent difference between HSPF and HSPF2 is about 19.0%.
![Develop](https://github.com/user-attachments/assets/a3d2d2a4-869c-4522-a8e8-953010153dcd)

Branch: percent difference between HSPF and HSPF2 is about 8.9%.
![Branch](https://github.com/user-attachments/assets/6955537d-b48e-48ef-9af6-5c3e64de86ff)

Nigusse commented:
See below comparison of the HSPF2 calculation intermediate outputs for develop and this branch:

Develop:
![Develop](https://github.com/user-attachments/assets/a72fe05e-bb5f-4fb0-a5e5-085d1a1d1717)

Branch:
![Branch](https://github.com/user-attachments/assets/61259719-6580-40e4-b1fd-b84b96759727)

The delivered heating rate is 4979.8W, the same for development and this branch. However, the total electric power input (compressor electric power + electric resistance heater power) decreased to 2389.0W from 2690.4W, resulting in an increased HSPF2 value. Note that the electric resistance heater power input has decreased more than the compressor electric power input increase.

Nigusse commented:
Attached is a test file used to generate sample results demonstrating the fix.
 
[Issue9909_in.idf.txt](https://github.com/user-attachments/files/16378355/Issue9909_in.idf.txt)

Nigusse committed:
Corrected equation re-write mistake

Nigusse committed:
Merge branch 'develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

Nigusse commented:
The regression test results show that 18 example files have diffs in the HSPF2 values. All these 18 example files have Multispeed DX Heating Coils. We found only six unique HSPF2 values since the inputs are similar in multiple example files. Below is a summary of the results based on the six files. The absolute diffs range 0.07 – 0.4 Btu/W-h, and the percent diffs range 1.0% - 8.1%. All the 18 example files with Multispeed DX heating coils in the branch showed increased HSPF2 values. However, they are less than the HSPF values. These results are consistent with the expectations.

![image](https://github.com/user-attachments/assets/cd008b86-7ac2-40b0-9624-0e5bc8f9b605)

Nigusse commented:
This branch is ready for review.

Myoldmopar reviewed (approved):
This is fine to me, and I know you've done your due diligence in these changes.  I'd _like_ another review opinion on the changes to speed lookups and such, but if not, it's OK.

Nigusse committed:
Merge branch 'develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

Nigusse committed:
Updated the new unit tests due to change in develop

RKStrand committed:
10634 Fix of Condenser Operation Issue

This commit includes the fix to the problem for both condensers operating even when the operation scheme said that one of them should not be operating.  It fixes this for all of the cooling tower models that were not set up to look at this.  Results now should that the towers are operating when they are supposed to.

RKStrand committed:
10634 Address Custom Check Error

Something about the logical flag wasn't what the custom check wanted.  This is an attempt to fix that.

RKStrand committed:
10634 Addition of Unit Test

Unit test was added to exercise the new subroutine that is being used to allow condensers to factor in condenser loop operation schemes.

RKStrand committed:
10634 Mass Flow to Zero

Corrected an issue with the fix where the mass flow rate when the load was zero was still positive in the defect idf when it should be zero.  This is now corrected.

RKStrand committed:
10634 Addresses Review Comments

Various review comments were received.  These were all addressed in this commit.

EnergyArchmage committed:
add SafeCopyPlantNode() to cooling towers

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

Myoldmopar committed:
Merge develop, cleanup unused vars

Myoldmopar commented:
@Nigusse I was going to merge this in, but during local testing I found 3 tests failing.  I spent about 30 minutes trying to diagnose the issue, looking at other changes made recently that might interact, but couldn't solve it quick enough.  While doing that I also found like 20+ variables in SizeDXCoil that were simply unused.  (Assigned but never accessed)  So I removed them just because.

If these 3 tests still fail, and you are able to look at them, that would be great.  If not, I'll have to do some further testing to figure out what I need to do to the unit tests to get them to work.  

Extra info:
 - This is the obvious culprit, but I couldn't tell quick enough what was needed on the HSPF tests.  
 - If I need to debug further, I'm likely going to roll back to the unmerged version of this branch to get it to pass, then more carefully bring in commits from develop to see exactly what is triggering the failures.

Anyway, I'm going to move on to other PRs for now.

Nigusse commented:
@Myoldmopar I will look at it.

Nigusse commented:
@Myoldmopar I will update the branch. If the failed unit tests are the new unit tests added, then I know what the problem is. I will look at it later today.

Nigusse commented:
@Myoldmopar The failed unit test must have impacted by changes somewhere else. Now, the heating capacity is sized to zero. Maybe these unit tests need to be modified.

Nigusse committed:
fixed field index

Nigusse committed:
Modified the failed unit tests by adding sizing parameters

Nigusse committed:
Format cleanup

Nigusse commented:
@Myoldmopar I modified the failed unit tests and all pass now. I ran the unit tests serially and all happy. This is good to go, albeit the format issue.

Nigusse committed:
Merge branch 'develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

Nigusse commented:
@Myoldmopar The failed unit tests and format warning message are addressed. This is ready for merging, I think.

Myoldmopar commented:
@Nigusse I'm going to have to figure out the last item here.  One test file is appearing to fail or something on Mac only.  I can't reproduce on Linux.  I'll take a look.

Myoldmopar commented:
I built on Mac and everything is still happy.  I'm hoping it's a fluke and I triggered a Mac rebuild.  Let's see.

Nigusse commented:
@Myoldmopar The Mac machine regression run has three additional example files showing diffs. I am not expecting any diffs in the following IDFs because changes in this branch do not impact them.

```
PurchAirWindowBlind_BlockBeamSolar.idf
PurchAirWindowBlind2.idf
PurchAirWindowBlind3.idf
```

rraustad committed:
Reduce scope

rraustad committed:
Shadow variable

rraustad committed:
Unused var

rraustad committed:
Unused and const

rraustad committed:
Unused and reduce scope

rraustad committed:
Reassigned OK to remove?

rraustad committed:
Unused/reduce

rraustad committed:
Unused and reduce scope

rraustad committed:
Assigned value before used

rraustad committed:
REduce scope HR

rraustad committed:
Reduce scope and shadow

rraustad committed:
Reduce scope

rraustad committed:
Reduce scope

rraustad committed:
Reduce scope

rraustad committed:
Reduce scope

Myoldmopar commented:
I think CI has magically calmed itself down, so I'm pulling develop in to get some fresh results here.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

jmarrec committed:
Don't catch std::exception in Debug mode so we can catch array bounds error in debugger

cmake will properly define NDEBUG on msvc as well in release mode
cf https://gitlab.kitware.com/cmake/cmake/-/blob/1e35163a/Modules/Platform/Windows-MSVC.cmake#L481-483

LipingWang committed:
Bug fix + Output Variable Reference

LipingWang committed:
Bug fix

rraustad committed:
First few logic fixes

rraustad committed:
AFN fixes

rraustad committed:
BranchInputManager

rraustad committed:
Chillers

rraustad committed:
const refs

rraustad committed:
Shadow vars

rraustad committed:
Const vars

rraustad committed:
Unused var

rraustad committed:
Const and shadow vars

rraustad committed:
eoh shadow

rraustad committed:
Shadow var Ctd

rraustad committed:
unused and unused outer shadow

rraustad committed:
Shadow var and unused

rraustad committed:
const vars

rraustad committed:
Const and shadow

rraustad committed:
all scope vars

LipingWang committed:
Update IO reference.

LipingWang committed:
format update

LipingWang committed:
output variable name

LipingWang committed:
update the csv & remove meters output

mjwitte committed:
Output variable transition and warning cleanup

rraustad committed:
Clean up some array references

rraustad committed:
Value never used

rraustad committed:
scope

rraustad committed:
reduce scope

rraustad committed:
Scope reduction

rraustad committed:
Scope reduction

rraustad committed:
const

rraustad committed:
scope

rraustad committed:
unused string

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
const, scope

rraustad committed:
scope

rraustad committed:
reduce

rraustad committed:
reduce scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
reduce

rraustad committed:
backup correct build error

rraustad committed:
scope

rraustad committed:
reduce

rraustad committed:
scope

rraustad committed:
reference

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
Run linux debug CI

rraustad committed:
Run Windows CI again

rraustad committed:
Clang format

rraustad committed:
conditionals and scope

rraustad committed:
scope

rraustad committed:
conditionals and scope

rraustad committed:
unused, constexpr and scope

rraustad committed:
scope and duplicate conditional

rraustad committed:
const and unused var

rraustad committed:
const

rraustad committed:
scope

rraustad committed:
Shadow var

rraustad committed:
scope

rraustad committed:
unused and shadow var

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
Shadow var and scope

rraustad committed:
scope

rraustad committed:
Value never used

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
clang fornat

rraustad committed:
redundant logic

rraustad committed:
const and reduce

rraustad committed:
const and unused

rraustad committed:
const, scope and shadow outer function

rraustad committed:
scope

rraustad committed:
scope

rraustad committed:
scope and const

rraustad committed:
const

rraustad committed:
Move to match previous code

Myoldmopar committed:
Remove third party nested gtest folder

Nigusse committed:
Code changed missed during resolving merge conflict

Nigusse committed:
Merge branch 'develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

Nigusse commented:
@Myoldmopar Resolved the conflict and the 18 example files diffs are expected per the original fix.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into Issue9909_multispeed_DX_Htg_Coil_HSPF2

mjwitte commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -7549,22 +7487,19 @@ void SizeDXCoil(EnergyPlusData &state, int const DXCoilNum)
             PrintFlag = true;
             state.dataSize->DataTotCapCurveIndex = thisDXCoil.CCapFTemp(Mode);
             if (thisDXCoil.DXCoilType_Num == HVAC::CoilDX_CoolingTwoStageWHumControl) {
-                SizingMethod = HVAC::CoolingCapacitySizing;
```
@Nigusse @rraustad Are these changes to DXCoils.cc supposed to be in this PR?

rraustad replied:
It looks like cleaning up unused information. I guess this can be done here as long as diffs show only for HSPF calculations.

Nigusse replied:
No. I did not make this change.

Nigusse replied:
> @Nigusse @rraustad Are these changes to DXCoils.cc supposed to be in this PR?

I think, @Myoldmopar might have made sizing related cleanup.

mjwitte replied:
Ah, ok, I see that commit.

Myoldmopar commented:
No conflicts here :tada: , but I'll still pull develop in locally to check nothing broke with the big refactor that just merged.

Myoldmopar commented:
(
And if you ever see a failure like that

![image](https://github.com/user-attachments/assets/5442dd10-c46e-4553-ae81-37b3e13c8c88)

You should be able to just click the Details tab and choose to re run the Jobs on the right side:

![image](https://github.com/user-attachments/assets/512116e4-c22a-4983-8f06-f9d7d06359cb)
)

Anyway, pulled and it's all happy, merging this in.  Thanks for your patience @Nigusse
